File: ios/BuzzTerminal/DesignSystem/Glyphs.swift
import SwiftUI

/// The four line glyphs the design draws inline as SVG.
///
/// Drawn as SwiftUI `Shape`s rather than shipped as assets or SF Symbols for one
/// reason that matters and one that is a bonus. The reason: SF Symbols has no
/// NFC-wave mark, and the design's is a specific three-arc form. The bonus:
/// stroke weight stays under our control, and Modernist cares about stroke
/// weight — the marks are 1.1–3.0 units in a 24-unit box depending on context.
///
/// Coordinates are transcribed from the design's `viewBox="0 0 24 24"` paths, so
/// the two can be diffed by eye.
enum SBGlyph {
    case nfcWave
    case check
    case blocked
    case alert
    case backspace

    /// Stroke width in viewBox units, as authored in the design.
    var authoredStrokeWidth: CGFloat {
        switch self {
        case .nfcWave: 1.2
        case .check: 2.4
        case .blocked: 2.4
        case .alert: 2.4
        case .backspace: 2.0
        }
    }
}

/// Renders an `SBGlyph` at a point size, scaling the stroke the way SVG does.
struct SBGlyphView: View {
    let glyph: SBGlyph
    var size: CGFloat
    var strokeWidth: CGFloat?
    var color: Color = .sbAccent

    var body: some View {
        let scale = size / 24
        let width = (strokeWidth ?? glyph.authoredStrokeWidth) * scale
        GlyphShape(glyph: glyph)
            .stroke(
                color,
                style: StrokeStyle(lineWidth: width, lineCap: .round, lineJoin: .round)
            )
            .frame(width: size, height: size)
    }
}

private struct GlyphShape: Shape {
    let glyph: SBGlyph

    func path(in rect: CGRect) -> Path {
        let s = min(rect.width, rect.height) / 24
        func p(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: rect.minX + x * s, y: rect.minY + y * s)
        }

        var path = Path()
        switch glyph {
        case .nfcWave:
            // Three nested arcs, widest on the left — the radio-wave mark.
            // <path d="M6 5c3.2 2.4 3.2 11.6 0 14"/> and two smaller siblings.
            path.move(to: p(6, 5))
            path.addCurve(to: p(6, 19), control1: p(9.2, 7.4), control2: p(9.2, 16.6))
            path.move(to: p(10.5, 7))
            path.addCurve(to: p(10.5, 17), control1: p(12.7, 8.7), control2: p(12.7, 15.3))
            path.move(to: p(15, 9.2))
            path.addCurve(to: p(15, 14.8), control1: p(16.2, 10.2), control2: p(16.2, 13.8))

        case .check:
            // <path d="M4 12.5l5 5L20 6.5"/>
            path.move(to: p(4, 12.5))
            path.addLine(to: p(9, 17.5))
            path.addLine(to: p(20, 6.5))

        case .blocked:
            // A circle struck through — <circle r="8.5"/> + <path d="M6.5 17.5L17.5 6.5"/>
            path.addEllipse(in: CGRect(
                x: rect.minX + 3.5 * s, y: rect.minY + 3.5 * s,
                width: 17 * s, height: 17 * s
            ))
            path.move(to: p(6.5, 17.5))
            path.addLine(to: p(17.5, 6.5))

        case .alert:
            // A bar and a dot — <path d="M12 6.5v8"/> + <path d="M12 18h.01"/>
            path.move(to: p(12, 6.5))
            path.addLine(to: p(12, 14.5))
            path.move(to: p(12, 18))
            path.addLine(to: p(12.01, 18))

        case .backspace:
            // Lucide's `delete`, with its 2-unit corner radii squared off —
            // Modernist does not round a corner anywhere, including in an icon.
            path.move(to: p(2, 12))
            path.addLine(to: p(9, 5))
            path.addLine(to: p(22, 5))
            path.addLine(to: p(22, 19))
            path.addLine(to: p(9, 19))
            path.closeSubpath()
            // The X inside.
            path.move(to: p(12, 9))
            path.addLine(to: p(18, 15))
            path.move(to: p(18, 9))
            path.addLine(to: p(12, 15))
        }
        return path
    }
}

#Preview("Glyphs") {
    HStack(spacing: SBSpace.x6) {
        SBGlyphView(glyph: .nfcWave, size: 46)
        SBGlyphView(glyph: .check, size: 32, color: .sbOk)
        SBGlyphView(glyph: .blocked, size: 32)
        SBGlyphView(glyph: .alert, size: 32)
    }
    .padding(SBSpace.x6)
    .background(Color.sbBackground)
}

File: android/app/src/main/kotlin/fest/swingbuzz/terminal/designsystem/Glyphs.kt
package fest.swingbuzz.terminal.designsystem

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * The five line glyphs the design draws inline as SVG.
 *
 * Drawn as paths on a [Canvas] rather than shipped as vector drawables for the
 * same reason the iOS side draws them as `Shape`s: stroke weight stays under our
 * control, and Modernist cares about stroke weight — the marks are 1.1–3.0 units
 * in a 24-unit box depending on context.
 *
 * Coordinates are transcribed from the design's `viewBox="0 0 24 24"` paths, so
 * this file and `ios/BuzzTerminal/DesignSystem/Glyphs.swift` can be diffed by
 * eye.
 */
enum class SBGlyph(
    /** Stroke width in viewBox units, as authored in the design. */
    val authoredStrokeWidth: Float,
) {
    NFC_WAVE(1.2f),
    CHECK(2.4f),
    BLOCKED(2.4f),
    ALERT(2.4f),
    BACKSPACE(2.0f),
}

/** Renders an [SBGlyph] at a size in dp, scaling the stroke the way SVG does. */
@Composable
fun SBGlyphView(
    glyph: SBGlyph,
    size: Dp,
    modifier: Modifier = Modifier,
    strokeWidth: Float? = null,
    color: Color = SB.accent,
) {
    Canvas(modifier.size(size)) {
        val s = minOf(this.size.width, this.size.height) / 24f
        drawPath(
            path = glyph.path(s),
            color = color,
            style = Stroke(
                width = (strokeWidth ?: glyph.authoredStrokeWidth) * s,
                cap = StrokeCap.Round,
                join = StrokeJoin.Round,
                pathEffect = PathEffect.cornerPathEffect(0f),
            ),
        )
    }
}

private fun SBGlyph.path(s: Float): Path {
    fun x(v: Float) = v * s
    fun y(v: Float) = v * s

    val path = Path()
    when (this) {
        SBGlyph.NFC_WAVE -> {
            // Three nested arcs, widest on the left — the radio-wave mark.
            // <path d="M6 5c3.2 2.4 3.2 11.6 0 14"/> and two smaller siblings.
            path.moveTo(x(6f), y(5f))
            path.cubicTo(x(9.2f), y(7.4f), x(9.2f), y(16.6f), x(6f), y(19f))
            path.moveTo(x(10.5f), y(7f))
            path.cubicTo(x(12.7f), y(8.7f), x(12.7f), y(15.3f), x(10.5f), y(17f))
            path.moveTo(x(15f), y(9.2f))
            path.cubicTo(x(16.2f), y(10.2f), x(16.2f), y(13.8f), x(15f), y(14.8f))
        }

        SBGlyph.CHECK -> {
            // <path d="M4 12.5l5 5L20 6.5"/>
            path.moveTo(x(4f), y(12.5f))
            path.lineTo(x(9f), y(17.5f))
            path.lineTo(x(20f), y(6.5f))
        }

        SBGlyph.BLOCKED -> {
            // A circle struck through — <circle r="8.5"/> + <path d="M6.5 17.5L17.5 6.5"/>
            path.addOval(Rect(Offset(x(3.5f), y(3.5f)), Size(x(17f), y(17f))))
            path.moveTo(x(6.5f), y(17.5f))
            path.lineTo(x(17.5f), y(6.5f))
        }

        SBGlyph.ALERT -> {
            // A bar and a dot — <path d="M12 6.5v8"/> + <path d="M12 18h.01"/>
            path.moveTo(x(12f), y(6.5f))
            path.lineTo(x(12f), y(14.5f))
            path.moveTo(x(12f), y(18f))
            path.lineTo(x(12.01f), y(18f))
        }

        SBGlyph.BACKSPACE -> {
            // Lucide's `delete`, with its 2-unit corner radii squared off —
            // Modernist does not round a corner anywhere, including in an icon.
            path.moveTo(x(2f), y(12f))
            path.lineTo(x(9f), y(5f))
            path.lineTo(x(22f), y(5f))
            path.lineTo(x(22f), y(19f))
            path.lineTo(x(9f), y(19f))
            path.close()
            // The X inside.
            path.moveTo(x(12f), y(9f))
            path.lineTo(x(18f), y(15f))
            path.moveTo(x(18f), y(9f))
            path.lineTo(x(12f), y(15f))
        }
    }
    return path
}

@Preview(showBackground = true, backgroundColor = 0xFFF3F2F2)
@Composable
private fun GlyphsPreview() {
    Row(
        modifier = Modifier.padding(SBSpace.x6),
        horizontalArrangement = Arrangement.spacedBy(SBSpace.x6),
    ) {
        SBGlyphView(SBGlyph.NFC_WAVE, 46.dp)
        SBGlyphView(SBGlyph.CHECK, 32.dp, color = SB.ok)
        SBGlyphView(SBGlyph.BLOCKED, 32.dp)
        SBGlyphView(SBGlyph.ALERT, 32.dp)
        SBGlyphView(SBGlyph.BACKSPACE, 32.dp, color = SB.ink)
    }
}
